File: Source/[Tools]/[Libraries]/Tools.Pdb/PdbFile.cpp
#include "Pdb.h"
#include "PdbFile.h"

using namespace Tools::Pdb;

PdbFile::PdbFile(String^ path)
{
    this->path = path;

    if (!File::Exists(path))
        throw gcnew IOException();

	HMODULE hModule = LoadLibrary(L"C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\Community\\DIA SDK\\bin\\msdia140.dll");
	if (!hModule)
		throw gcnew Exception();

	BOOL(WINAPI* DllGetClassObject)(REFCLSID, REFIID, LPVOID*) = (BOOL(WINAPI*)(REFCLSID, REFIID, LPVOID*))GetProcAddress(hModule, "DllGetClassObject");
	if (!DllGetClassObject)
		throw gcnew Exception();

	CComPtr<IClassFactory> pClassFactory;
	HRESULT hResult = DllGetClassObject(CLSID_DiaSource, IID_IClassFactory, (LPVOID*)&pClassFactory);
	if (FAILED(hResult))
		throw gcnew Exception();

	IDiaDataSource* diaDataSourcePointer = NULL;

	hResult = pClassFactory->CreateInstance(NULL, IID_IDiaDataSource, (void**)&diaDataSourcePointer);
	if (FAILED(hResult))
		throw gcnew Exception();

	diaDataSource = diaDataSourcePointer;

    LPCOLESTR pathStr = (LPCOLESTR)(Marshal::StringToHGlobalUni(path).ToPointer());
    diaDataSource->loadDataFromPdb(pathStr);
}

PdbSession^ PdbFile::OpenSession(ULONGLONG address)
{
    IDiaSession* diaSession = NULL;
    diaDataSource->openSession(&diaSession);

    diaSession->put_loadAddress(address);

    return gcnew PdbSession(diaSession);
}

File: Source/[Tools]/[Libraries]/Tools.Pdb/PdbFile.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Dia2Lib;

namespace Tools.Pdb
{
    public class PdbFile
    {
        private IDiaDataSource diaDataSource;

        public PdbFile(string path)
        {
            if (!File.Exists(path))
                throw new FileNotFoundException(path);

            diaDataSource = new DiaSource();
            diaDataSource.loadDataFromPdb(path);
        }

        public PdbSession OpenSession(ulong address)
        {
            diaDataSource.openSession(out IDiaSession diaSession);
            diaSession.loadAddress = address;
            return new PdbSession(diaSession);
        }
    }
}
